File: server/src/commands/user/memberlist.py
from .. import register_command

import socket
import requests

from src.colors import *
from src.db import Database

from init import User, ClientSender, server_dir, users, config, online_mode, global_ip, max_users
from src.vars import api
from src.functions import memberListNickname, memberListBadge, isOnline, check_user_status

import yaml
from yaml import SafeLoader


@register_command("memberlist")
@register_command("userlist")
def memberlist_command(socket: socket.socket, user: User, args: list, sender: ClientSender):
    badge = {}
    
    cmd_db = Database(server_dir + "/users.db", check_same_thread=False)
                
    cmd_db.execute("SELECT username FROM users")
    raw_members = cmd_db.fetchall()
    membersLen  = len([raw_members for raw_members in sorted(raw_members)])
    
    cmd_db.execute("SELECT username, badge FROM users WHERE role = 'admin'")
    raw_admins  = cmd_db.fetchall()
    admins_len  = len([raw_admins for raw_admins in sorted(raw_admins)])
    admins      = "\n           ".join([f"{check_user_status(type='name', user=result[0])}{LIGHTRED_EX} {memberListNickname(result[0])} {memberListBadge(result[0])}" for result in raw_admins])
    
    cmd_db.execute("SELECT username, badge FROM users WHERE role = 'bot'")
    raw_bots    = cmd_db.fetchall()
    bots_len    = len([raw_bots for raw_bots in sorted(raw_bots)])
    bots        = "\n           ".join([f"{check_user_status(type='name', user=result[0])}{LIGHTMAGENTA_EX} {memberListNickname(result[0])} {memberListBadge(result[0])}" for result in raw_bots])
    
    
    cmd_db.execute("SELECT username, badge FROM users WHERE role = 'member'")
    raw_members = cmd_db.fetchall()
    members_len = len([raw_members for raw_members in sorted(raw_members)])
    members     = "\n           ".join([f"{check_user_status(type='name', user=result[0])}{LIGHTYELLOW_EX} {memberListNickname(result[0])} {memberListBadge(result[0])}" for result in raw_members])
    
    try:
        if online_mode == True:
            verified = requests.get(api + "server/verified?addr=" + global_ip)
        
            if verified.text == "True":
                verified_txt = f"{GREEN}[VERIFIED]{CYAN} "
            else:
                verified_txt = ""
            
        else:
            verified_txt = ""
        
    except Exception as e: 
        print(e)
    
    onlineUsersLen2 = len([user for user in sorted(users.values())])
    
    _online_users = f"{onlineUsersLen2}/{max_users}"
    if max_users == -1: _online_users = f"{onlineUsersLen2}"
                
    sender.send(f"""{CYAN +  Colors.UNDERLINE + Colors.BOLD}{verified_txt}{config['server']['name'].upper()} ({membersLen} Members, {_online_users} Online){RESET + Colors.RESET}
        {Colors.BOLD}->{Colors.RESET} {RED}Administrators ({admins_len}){RESET}
           {LIGHTRED_EX}{admins}{RESET}
        
        {Colors.BOLD}->{Colors.RESET} {MAGENTA}Bots ({bots_len}){RESET}
           {LIGHTMAGENTA_EX}{bots}{RESET}
    
        {Colors.BOLD}->{Colors.RESET} {YELLOW}Members ({members_len}){RESET}
           {LIGHTYELLOW_EX}{members}{RESET}
    """
    
    )

File: src/commands/user/memberlist.rs
use stblib::colors::{BOLD, C_RESET, CYAN, RED, UNDERLINE, RESET, MAGENTA, YELLOW, LIGHT_YELLOW, LIGHT_MAGENTA, LIGHT_RED};

use crate::system_core::commands;
use crate::system_core::commands::CommandCategory;
use crate::system_core::internals::MessageToClient;
use crate::system_core::permissions::Permissions;
use crate::system_core::server_core::{get_online_usernames, STATUS};
use crate::database::DATABASE;
use crate::utilities::parse_user_status;

async fn format_to_list(member_list: &[String], fmt_color: &str) -> String {
    let futures = member_list.iter().map(|result| {
        async {
            let user = DATABASE.get_user_by_name(result).await.unwrap();

            let badge = if user.badge.is_empty() {
                String::new()
            } else {
                format!("[{}]", user.badge)
            };


            let username = if user.nickname.is_empty() {
                String::from(&user.username)
            } else {
                format!("{} (@{})", user.nickname, user.username)
            };

            let status_raw = *STATUS.read().await.get_by_name(user.username.as_str());
            let status_symbol = parse_user_status(status_raw).0;

            format!("{C_RESET}{status_symbol}{fmt_color} {username} {badge}")
        }
    });

    let formatted_results = futures::future::join_all(futures).await;
    formatted_results.join("\n           ")
}


pub fn memberlist() -> commands::Command {
    async fn logic(ctx: &commands::Context) -> commands::CommandResponse {
        let raw_members = DATABASE.get_members_by_role("member").await;

        let members: String = format_to_list(&raw_members, LIGHT_YELLOW).await;

        let raw_bots: Vec<String> = DATABASE.get_members_by_role("bot").await;

        let bots: String = format_to_list(&raw_bots, LIGHT_MAGENTA).await;

        let raw_admins: Vec<String> = DATABASE.get_members_by_role("admin").await;

        let admins: String = format_to_list(&raw_admins, LIGHT_RED).await;

        let members_total = DATABASE.get_members().await.len();
        let online_users = get_online_usernames().await.len();


        let message = format!(
            "
        {CYAN}{UNDERLINE}({members_total} Members, {online_users} Online){C_RESET}
        {BOLD}->{C_RESET} {RED}Administrators ({}){RESET}
           {LIGHT_RED}{admins}{RESET}

        {BOLD}->{RESET} {MAGENTA}Bots ({}){RESET}
           {LIGHT_MAGENTA}{bots}{RESET}

        {BOLD}->{C_RESET} {YELLOW}Members ({}){RESET}
           {LIGHT_YELLOW}{members}{RESET}",
            raw_admins.len(), raw_bots.len(), raw_members.len()

        );


        ctx.tx_channel.send(MessageToClient::SystemMessage {
            content: message
        }).await.unwrap();

        Ok(None)
    }

    commands::Command {
        name: "memberlist".to_string(),
        aliases: vec!["userlist", "users"],
        description: "Displays a list of members with their badges and roles".to_string(),
        category: CommandCategory::User,
        permissions: Permissions::Member,
        required_args: 0,
        handler: |ctx| Box::pin(async move {
            logic(&ctx).await
        }),
    }
}